File: app/routes/provider.py
"""API 平台与全局配置：/api/providers/*, /api/config*, /api/models。"""

import json
import os
import re
from typing import List

import httpx
from fastapi import APIRouter, HTTPException

from .. import config, providers
from ..models import ApiProviderPayload, TestConnectionPayload

router = APIRouter()


@router.get("/api/config")
async def ai_config():
    preferred_chat_model = next(
        (m for m in config.CHAT_MODELS if m == "gpt-5.5"),
        config.CHAT_MODELS[0] if config.CHAT_MODELS else config.CHAT_MODEL,
    )
    public = [providers.public_provider(p) for p in providers.load_api_providers()]
    return {
        "base_url": config.AI_BASE_URL,
        "chat_model": preferred_chat_model,
        "image_model": config.IMAGE_MODEL,
        "chat_models": config.CHAT_MODELS,
        "image_models": config.IMAGE_MODELS,
        "video_models": config.VIDEO_MODELS,
        "api_providers": public,
        "has_api_key": bool(config.AI_API_KEY),
        "ms_chat_models": config.MODELSCOPE_CHAT_MODELS,
        "has_ms_key": bool(config.MODELSCOPE_API_KEY),
    }


@router.get("/api/models")
async def ai_models():
    return {
        "chat_models": config.CHAT_MODELS,
        "image_models": config.IMAGE_MODELS,
        "video_models": config.VIDEO_MODELS,
    }


@router.get("/api/providers")
async def api_providers_get():
    return {"providers": [providers.public_provider(p) for p in providers.load_api_providers()]}


@router.put("/api/providers")
async def save_providers(payload: List[ApiProviderPayload]):
    normalized = []
    env_updates = {}
    raw_primary_flags = [bool(getattr(item, "primary", False)) for item in payload]
    for item in payload:
        provider = providers.normalize_provider(item.dict(exclude={"api_key"}))
        if any(existing["id"] == provider["id"] for existing in normalized):
            raise HTTPException(status_code=400, detail=f"API 平台 ID 重复：{provider['id']}")
        normalized.append(provider)
        if item.api_key is not None:
            env_updates[providers.provider_key_env(provider["id"])] = item.api_key.strip()
        if provider["id"] == "comfly":
            env_updates["COMFLY_BASE_URL"] = provider["base_url"]
            env_updates["IMAGE_MODELS"] = ",".join(provider["image_models"])
            env_updates["CHAT_MODELS"] = ",".join(provider["chat_models"])
            env_updates["VIDEO_MODELS"] = ",".join(provider.get("video_models") or [])
        if provider["id"] == "modelscope":
            env_updates["MODELSCOPE_CHAT_MODELS"] = ",".join(provider["chat_models"])
    if not normalized:
        raise HTTPException(status_code=400, detail="至少保留一个 API 平台")
    # 最多一个 primary：取最后被标记的；都没标记则保持原样不强制
    primary_indices = [i for i, flag in enumerate(raw_primary_flags) if flag]
    if primary_indices:
        winner = primary_indices[-1]
        for i, p in enumerate(normalized):
            p["primary"] = (i == winner)
    providers.save_api_providers(normalized)
    if env_updates:
        providers.update_env_values(env_updates)
        config.reload_env_globals()  # 立即把最新 env 同步回模块全局变量
    return {"providers": [providers.public_provider(p) for p in normalized]}


@router.get("/api/config/token")
async def get_global_token():
    """ModelScope token 从 env 读取（不再支持 UI 修改），兼容旧版 global_config.json。"""
    if config.MODELSCOPE_API_KEY:
        return {"token": config.MODELSCOPE_API_KEY}
    if os.path.exists(config.GLOBAL_CONFIG_FILE):
        try:
            with open(config.GLOBAL_CONFIG_FILE, "r", encoding="utf-8") as f:
                cfg = json.load(f)
                return {"token": cfg.get("modelscope_token", "")}
        except Exception:
            pass
    return {"token": ""}


def _classify_model_id(mid: str) -> str:
    lc = mid.lower()
    video_keys = ["veo", "sora", "wan2", "wanx", "doubao-seedance", "doubao-1", "kling", "hailuo", "video", "t2v-", "i2v-", "s2v"]
    if any(k in lc for k in video_keys):
        return "video"
    image_keys = ["image", "dalle", "dall-e", "imagen", "flux", "stable", "sdxl", "midjourney", "nano-banana", "ideogram", "fal-ai", "z-image", "qwen-image", "klein"]
    if any(k in lc for k in image_keys):
        return "image"
    return "chat"


def _extract_model_ids(raw) -> list:
    items = raw.get("data") if isinstance(raw, dict) else None
    if not items and isinstance(raw, dict):
        items = raw.get("models") or raw.get("list") or []
    if not isinstance(items, list):
        items = []
    ids = []
    for it in items:
        if isinstance(it, str):
            ids.append(it)
        elif isinstance(it, dict):
            mid = it.get("id") or it.get("name") or it.get("model")
            if mid:
                ids.append(str(mid))
    return sorted(set(ids))


@router.post("/api/providers/test-connection")
async def test_provider_connection(payload: TestConnectionPayload):
    """测试请求地址：调上游 /v1/models；通过则同时返回按类别分组的模型清单。"""
    base_url = (payload.base_url or "").strip().rstrip("/")
    if not base_url:
        raise HTTPException(status_code=400, detail="请先填写请求地址")
    if not re.match(r"^https?://", base_url):
        raise HTTPException(status_code=400, detail="请求地址必须以 http:// 或 https:// 开头")
    api_key = (payload.api_key or "").strip()
    if not api_key and payload.provider_id:
        api_key = os.getenv(providers.provider_key_env(payload.provider_id), "")
    if not api_key:
        raise HTTPException(status_code=400, detail="请先填写或保存 API Key")
    url = f"{base_url}/models" if base_url.endswith("/v1") else f"{base_url}/v1/models"
    try:
        async with httpx.AsyncClient(timeout=15) as client:
            resp = await client.get(url, headers={"Authorization": f"Bearer {api_key}", "Accept": "application/json"})
        if resp.status_code >= 400:
            return {"ok": False, "status": resp.status_code, "message": resp.text[:300]}
        data = resp.json() if resp.text else {}
        ids = _extract_model_ids(data)
        grouped = {"image": [], "chat": [], "video": []}
        for mid in ids:
            grouped[_classify_model_id(mid)].append(mid)
        return {
            "ok": True,
            "status": resp.status_code,
            "model_count": len(ids),
            "image_models": grouped["image"],
            "chat_models": grouped["chat"],
            "video_models": grouped["video"],
            "all": ids,
        }
    except httpx.HTTPError as e:
        return {"ok": False, "status": 0, "message": str(e)[:300]}


@router.post("/api/providers/probe-async")
async def probe_async_endpoint(payload: TestConnectionPayload):
    """验证异步协议：用假 task_id 请求 GET /v1/tasks/{fake_id}。

    400 "Invalid task ID" → 端点存在且 Key 有效；
    401/403 → Key 无效；
    404/连接失败 → 不支持异步端点。"""
    base_url = (payload.base_url or "").strip().rstrip("/")
    if not base_url:
        raise HTTPException(status_code=400, detail="请先填写请求地址")
    api_key = (payload.api_key or "").strip()
    if not api_key and payload.provider_id:
        api_key = os.getenv(providers.provider_key_env(payload.provider_id), "")
    if not api_key:
        raise HTTPException(status_code=400, detail="请先填写或保存 API Key")
    tasks_base = base_url if base_url.endswith("/v1") else f"{base_url}/v1"
    probe_url = f"{tasks_base}/tasks/healthcheck_probe_do_not_submit"
    try:
        async with httpx.AsyncClient(timeout=15) as client:
            resp = await client.get(probe_url, headers={"Authorization": f"Bearer {api_key}", "Accept": "application/json"})
        try:
            body = resp.json()
        except Exception:
            body = resp.text[:500]
        sc = resp.status_code
        err_msg = ""
        if isinstance(body, dict):
            err = body.get("error") or {}
            if isinstance(err, dict):
                err_msg = str(err.get("message") or "").lower()
            else:
                err_msg = str(err).lower()
        if sc == 400 and "invalid task id" in err_msg:
            return {"ok": True, "status_code": sc, "message": "异步任务端点可用，API Key 已通过认证", "raw": body}
        if sc in (401, 403):
            return {"ok": False, "status_code": sc, "message": "API Key 无效或无权限", "raw": body}
        if sc == 404:
            return {"ok": False, "status_code": sc, "message": "平台不支持 /v1/tasks/ 端点，可能不是 APIMart 异步协议", "raw": body}
        if 400 <= sc < 500:
            return {"ok": None, "status_code": sc, "message": f"端点返回 {sc}，请查看原始响应判断", "raw": body}
        if sc < 300:
            return {"ok": True, "status_code": sc, "message": f"端点返回 {sc}（意外成功）", "raw": body}
        return {"ok": False, "status_code": sc, "message": f"服务端错误 {sc}", "raw": body}
    except httpx.HTTPError as e:
        raise HTTPException(status_code=502, detail=str(e)[:300])


@router.get("/api/providers/{provider_id}/fetch-models")
async def fetch_upstream_models(provider_id: str):
    """从上游 OpenAI 兼容接口拉取 /v1/models 列表，按名称智能分类为 image/chat/video。"""
    provider = providers.get_api_provider(provider_id)
    base_url = (provider.get("base_url") or "").rstrip("/")
    if not base_url:
        raise HTTPException(status_code=400, detail=f"{provider.get('name') or provider_id} 未配置 Base URL")
    api_key = os.getenv(providers.provider_key_env(provider["id"]), "")
    if not api_key:
        raise HTTPException(status_code=400, detail=f"{provider.get('name') or provider_id} 未配置 API Key")
    url = f"{base_url}/models" if base_url.endswith("/v1") else f"{base_url}/v1/models"
    try:
        async with httpx.AsyncClient(timeout=30) as client:
            resp = await client.get(url, headers={"Authorization": f"Bearer {api_key}", "Accept": "application/json"})
            if resp.status_code >= 400:
                raise HTTPException(status_code=resp.status_code, detail=f"上游 /v1/models 失败：{resp.text[:300]}")
            raw = resp.json()
    except httpx.HTTPError as e:
        raise HTTPException(status_code=502, detail=f"请求上游模型列表失败：{e}")
    ids = _extract_model_ids(raw)
    grouped = {"image": [], "chat": [], "video": []}
    for mid in ids:
        grouped[_classify_model_id(mid)].append(mid)
    return {
        "total": len(ids),
        "image_models": grouped["image"],
        "chat_models": grouped["chat"],
        "video_models": grouped["video"],
        "all": ids,
    }

File: app-go/internal/handler/provider.go
package handler

import (
	"encoding/json"
	"io"
	"net/http"
	"os"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/gin-gonic/gin"

	"infinite-canvas/app-go/internal/store"
)

type APIProviderPayload struct {
	ID                      string           `json:"id"`
	Name                    string           `json:"name"`
	BaseURL                 string           `json:"base_url"`
	Protocol                string           `json:"protocol"`
	ImageGenerationEndpoint string           `json:"image_generation_endpoint"`
	ImageEditEndpoint       string           `json:"image_edit_endpoint"`
	Enabled                 bool             `json:"enabled"`
	Primary                 bool             `json:"primary"`
	ImageModels             []string         `json:"image_models"`
	ChatModels              []string         `json:"chat_models"`
	VideoModels             []string         `json:"video_models"`
	MSLoras                 []map[string]any `json:"ms_loras"`
	MSDefaultsVersion       int              `json:"ms_defaults_version"`
	APIKey                  *string          `json:"api_key"`
	ClearKey                bool             `json:"clear_key"`
}

type FirstPartyKeysPayload struct {
	OpenAIAPIKey              *string `json:"openai_api_key"`
	VolcengineArkAPIKey       *string `json:"volcengine_ark_api_key"`
	VolcengineAccessKeyID     *string `json:"volcengine_access_key_id"`
	VolcengineSecretAccessKey *string `json:"volcengine_secret_access_key"`
	VolcengineProjectName     *string `json:"volcengine_project_name"`
	VolcengineRegion          *string `json:"volcengine_region"`
}

type TestConnectionPayload struct {
	BaseURL    string `json:"base_url"`
	APIKey     string `json:"api_key"`
	ProviderID string `json:"provider_id"`
}

func (h *Handler) GetConfig(c *gin.Context) {
	preferredChatModel := h.cfg.ChatModel
	if len(h.cfg.ChatModels) > 0 {
		preferredChatModel = h.cfg.ChatModels[0]
		for _, model := range h.cfg.ChatModels {
			if model == "gpt-5.5" {
				preferredChatModel = model
				break
			}
		}
	}

	c.JSON(http.StatusOK, gin.H{
		"base_url":                      h.cfg.AIBaseURL,
		"chat_model":                    preferredChatModel,
		"image_model":                   h.cfg.ImageModel,
		"chat_models":                   h.cfg.ChatModels,
		"image_models":                  h.cfg.ImageModels,
		"video_models":                  h.cfg.VideoModels,
		"api_providers":                 h.store.PublicAPIProviders(),
		"has_api_key":                   h.cfg.AIAPIKey != "",
		"ms_chat_models":                h.cfg.ModelScopeChatModels,
		"has_ms_key":                    h.cfg.ModelScopeAPIKey != "",
		"has_openai_key":                h.cfg.OpenAIAPIKey != "",
		"openai_key_preview":            store.MaskSecret(h.cfg.OpenAIAPIKey),
		"has_volcengine_ark_key":        h.cfg.VolcengineArkAPIKey != "",
		"volcengine_ark_key_preview":    store.MaskSecret(h.cfg.VolcengineArkAPIKey),
		"has_volcengine_access_key":     h.cfg.VolcengineAccessKeyID != "",
		"volcengine_access_key_preview": store.MaskSecret(h.cfg.VolcengineAccessKeyID),
		"has_volcengine_secret_key":     h.cfg.VolcengineSecretAccessKey != "",
		"volcengine_secret_key_preview": store.MaskSecret(h.cfg.VolcengineSecretAccessKey),
		"volcengine_project_name":       h.cfg.VolcengineProjectName,
		"volcengine_region":             h.cfg.VolcengineRegion,
	})
}

func (h *Handler) ListModels(c *gin.Context) {
	c.JSON(http.StatusOK, gin.H{
		"chat_models":  h.cfg.ChatModels,
		"image_models": h.cfg.ImageModels,
		"video_models": h.cfg.VideoModels,
	})
}

func (h *Handler) ListProviders(c *gin.Context) {
	c.JSON(http.StatusOK, gin.H{"providers": h.store.PublicAPIProviders()})
}

func (h *Handler) SaveProviders(c *gin.Context) {
	var payload []APIProviderPayload
	if err := c.ShouldBindJSON(&payload); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "请求参数不正确"})
		return
	}
	if len(payload) == 0 {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "至少保留一个 API 平台"})
		return
	}

	normalized := make([]map[string]any, 0, len(payload))
	envUpdates := map[string]string{}
	lastPrimary := -1
	for i, item := range payload {
		if item.Primary {
			lastPrimary = i
		}
		provider, err := store.NormalizeProviderForSave(map[string]any{
			"id":                        item.ID,
			"name":                      item.Name,
			"base_url":                  item.BaseURL,
			"protocol":                  item.Protocol,
			"image_generation_endpoint": item.ImageGenerationEndpoint,
			"image_edit_endpoint":       item.ImageEditEndpoint,
			"enabled":                   item.Enabled,
			"primary":                   item.Primary,
			"image_models":              item.ImageModels,
			"chat_models":               item.ChatModels,
			"video_models":              item.VideoModels,
			"ms_loras":                  item.MSLoras,
			"ms_defaults_version":       item.MSDefaultsVersion,
		})
		if err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"detail": err.Error()})
			return
		}
		for _, existing := range normalized {
			if existing["id"] == provider["id"] {
				c.JSON(http.StatusBadRequest, gin.H{"detail": "API 平台 ID 重复：" + stringValue(provider["id"], "")})
				return
			}
		}
		keyEnv := store.ProviderKeyEnv(stringValue(provider["id"], ""))
		if item.ClearKey {
			envUpdates[keyEnv] = ""
		} else if item.APIKey != nil && *item.APIKey != "" {
			envUpdates[keyEnv] = *item.APIKey
		}
		if provider["id"] == "comfly" {
			envUpdates["COMFLY_BASE_URL"] = stringValue(provider["base_url"], "")
			envUpdates["IMAGE_MODELS"] = joinStringAny(provider["image_models"])
			envUpdates["CHAT_MODELS"] = joinStringAny(provider["chat_models"])
			envUpdates["VIDEO_MODELS"] = joinStringAny(provider["video_models"])
		}
		if provider["id"] == "modelscope" {
			envUpdates["MODELSCOPE_CHAT_MODELS"] = joinStringAny(provider["chat_models"])
		}
		normalized = append(normalized, provider)
	}
	if lastPrimary >= 0 {
		for i := range normalized {
			normalized[i]["primary"] = i == lastPrimary
		}
	}

	if err := h.store.SaveAPIProviders(normalized); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"detail": "写入 API 平台配置失败：" + err.Error()})
		return
	}
	if err := h.store.UpdateEnvValues(envUpdates); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"detail": "写入 env 失败：" + err.Error()})
		return
	}
	h.cfg.RefreshFromEnv()
	c.JSON(http.StatusOK, gin.H{"providers": h.store.PublicAPIProviders()})
}

func (h *Handler) GetFirstPartyKeys(c *gin.Context) {
	c.JSON(http.StatusOK, h.firstPartyKeysResponse())
}

func (h *Handler) TestProviderConnection(c *gin.Context) {
	var payload TestConnectionPayload
	if err := c.ShouldBindJSON(&payload); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "请求参数不正确"})
		return
	}
	baseURL := strings.TrimRight(strings.TrimSpace(payload.BaseURL), "/")
	if baseURL == "" {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "请先填写请求地址"})
		return
	}
	if !strings.HasPrefix(baseURL, "http://") && !strings.HasPrefix(baseURL, "https://") {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "请求地址必须以 http:// 或 https:// 开头"})
		return
	}
	apiKey := strings.TrimSpace(payload.APIKey)
	if apiKey == "" && payload.ProviderID != "" {
		apiKey = os.Getenv(store.ProviderKeyEnv(payload.ProviderID))
	}
	if apiKey == "" {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "请先填写或保存 API Key"})
		return
	}
	modelURL := modelsEndpoint(baseURL)
	raw, status, err := getJSONWithBearer(modelURL, apiKey, 15*time.Second)
	if err != nil {
		c.JSON(http.StatusOK, gin.H{"ok": false, "status": 0, "message": truncate(err.Error(), 300)})
		return
	}
	if status >= 400 {
		c.JSON(http.StatusOK, gin.H{"ok": false, "status": status, "message": truncate(rawText(raw), 300)})
		return
	}
	ids := extractModelIDs(raw)
	grouped := groupModelIDs(ids)
	c.JSON(http.StatusOK, gin.H{
		"ok":           true,
		"status":       status,
		"model_count":  len(ids),
		"image_models": grouped["image"],
		"chat_models":  grouped["chat"],
		"video_models": grouped["video"],
		"all":          ids,
	})
}

func (h *Handler) FetchModelsFromPayload(c *gin.Context) {
	var payload TestConnectionPayload
	if err := c.ShouldBindJSON(&payload); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "请求参数不正确"})
		return
	}
	apiKey := strings.TrimSpace(payload.APIKey)
	if apiKey == "" && payload.ProviderID != "" {
		apiKey = os.Getenv(store.ProviderKeyEnv(payload.ProviderID))
	}
	h.fetchModelsFromUpstream(c, payload.BaseURL, apiKey)
}

func (h *Handler) FetchModelsFromSavedProvider(c *gin.Context) {
	providerID := c.Param("provider_id")
	provider, ok := h.providerByID(providerID)
	if !ok {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "未找到 API 平台：" + providerID})
		return
	}
	if enabled, ok := provider["enabled"].(bool); ok && !enabled {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "API 平台已禁用：" + stringValue(provider["name"], providerID)})
		return
	}
	apiKey := os.Getenv(store.ProviderKeyEnv(stringValue(provider["id"], providerID)))
	if apiKey == "" {
		c.JSON(http.StatusBadRequest, gin.H{"detail": stringValue(provider["name"], providerID) + " 未配置 API Key"})
		return
	}
	h.fetchModelsFromUpstream(c, stringValue(provider["base_url"], ""), apiKey)
}

func (h *Handler) ProbeAsyncEndpoint(c *gin.Context) {
	var payload TestConnectionPayload
	if err := c.ShouldBindJSON(&payload); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "请求参数不正确"})
		return
	}
	baseURL := strings.TrimRight(strings.TrimSpace(payload.BaseURL), "/")
	if baseURL == "" {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "请先填写请求地址"})
		return
	}
	apiKey := strings.TrimSpace(payload.APIKey)
	if apiKey == "" && payload.ProviderID != "" {
		apiKey = os.Getenv(store.ProviderKeyEnv(payload.ProviderID))
	}
	if apiKey == "" {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "请先填写或保存 API Key"})
		return
	}
	tasksBase := baseURL
	if !strings.HasSuffix(tasksBase, "/v1") {
		tasksBase += "/v1"
	}
	raw, status, err := getJSONWithBearer(tasksBase+"/tasks/healthcheck_probe_do_not_submit", apiKey, 15*time.Second)
	if err != nil {
		c.JSON(http.StatusBadGateway, gin.H{"detail": truncate(err.Error(), 300)})
		return
	}
	errMsg := ""
	if errObj, ok := raw["error"].(map[string]any); ok {
		errMsg = strings.ToLower(stringValue(errObj["message"], ""))
	} else if raw["error"] != nil {
		errMsg = strings.ToLower(stringValue(raw["error"], ""))
	}
	switch {
	case status == 400 && strings.Contains(errMsg, "invalid task id"):
		c.JSON(http.StatusOK, gin.H{"ok": true, "status_code": status, "message": "异步任务端点可用，API Key 已通过认证", "raw": raw})
	case status == 401 || status == 403:
		c.JSON(http.StatusOK, gin.H{"ok": false, "status_code": status, "message": "API Key 无效或无权限", "raw": raw})
	case status == 404:
		c.JSON(http.StatusOK, gin.H{"ok": false, "status_code": status, "message": "平台不支持 /v1/tasks/ 端点，可能不是 APIMart 异步协议", "raw": raw})
	case status >= 400 && status < 500:
		c.JSON(http.StatusOK, gin.H{"ok": nil, "status_code": status, "message": "端点返回 " + strconv.Itoa(status) + "，请查看原始响应判断", "raw": raw})
	case status < 300:
		c.JSON(http.StatusOK, gin.H{"ok": true, "status_code": status, "message": "端点返回 " + strconv.Itoa(status) + "（意外成功）", "raw": raw})
	default:
		c.JSON(http.StatusOK, gin.H{"ok": false, "status_code": status, "message": "服务端错误 " + strconv.Itoa(status), "raw": raw})
	}
}

func (h *Handler) SaveFirstPartyKeys(c *gin.Context) {
	var payload FirstPartyKeysPayload
	if err := c.ShouldBindJSON(&payload); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "请求参数不正确"})
		return
	}
	updates := map[string]string{}
	if payload.OpenAIAPIKey != nil {
		updates["OPENAI_API_KEY"] = *payload.OpenAIAPIKey
	}
	if payload.VolcengineArkAPIKey != nil {
		updates["VOLCENGINE_ARK_API_KEY"] = *payload.VolcengineArkAPIKey
	}
	if payload.VolcengineAccessKeyID != nil {
		updates["VOLCENGINE_ACCESS_KEY_ID"] = *payload.VolcengineAccessKeyID
	}
	if payload.VolcengineSecretAccessKey != nil {
		updates["VOLCENGINE_SECRET_ACCESS_KEY"] = *payload.VolcengineSecretAccessKey
	}
	if payload.VolcengineProjectName != nil {
		updates["VOLCENGINE_PROJECT_NAME"] = firstNonEmptyString(*payload.VolcengineProjectName, "default")
	}
	if payload.VolcengineRegion != nil {
		updates["VOLCENGINE_REGION"] = firstNonEmptyString(*payload.VolcengineRegion, "cn-beijing")
	}
	if err := h.store.UpdateEnvValues(updates); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"detail": "写入 env 失败：" + err.Error()})
		return
	}
	h.cfg.RefreshFromEnv()
	c.JSON(http.StatusOK, h.firstPartyKeysResponse())
}

func (h *Handler) GetGlobalToken(c *gin.Context) {
	if h.cfg.ModelScopeAPIKey != "" {
		c.JSON(http.StatusOK, gin.H{"token": h.cfg.ModelScopeAPIKey})
		return
	}
	data, err := os.ReadFile(h.cfg.GlobalConfigFile)
	if err == nil {
		var cfg map[string]any
		if json.Unmarshal(data, &cfg) == nil {
			c.JSON(http.StatusOK, gin.H{"token": stringValue(cfg["modelscope_token"], "")})
			return
		}
	}
	c.JSON(http.StatusOK, gin.H{"token": ""})
}

func (h *Handler) fetchModelsFromUpstream(c *gin.Context, baseURL, apiKey string) {
	baseURL = strings.TrimRight(strings.TrimSpace(baseURL), "/")
	if baseURL == "" {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "请先填写请求地址"})
		return
	}
	if !strings.HasPrefix(baseURL, "http://") && !strings.HasPrefix(baseURL, "https://") {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "请求地址必须以 http:// 或 https:// 开头"})
		return
	}
	if strings.TrimSpace(apiKey) == "" {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "请先填写或保存 API Key"})
		return
	}
	raw, status, err := getJSONWithBearer(modelsEndpoint(baseURL), apiKey, 30*time.Second)
	if err != nil {
		c.JSON(http.StatusBadGateway, gin.H{"detail": "请求上游模型列表失败：" + err.Error()})
		return
	}
	if status >= 400 {
		c.JSON(status, gin.H{"detail": "上游 /v1/models 失败：" + truncate(rawText(raw), 300)})
		return
	}
	ids := extractModelIDs(raw)
	grouped := groupModelIDs(ids)
	c.JSON(http.StatusOK, gin.H{
		"total":        len(ids),
		"image_models": grouped["image"],
		"chat_models":  grouped["chat"],
		"video_models": grouped["video"],
		"all":          ids,
	})
}

func (h *Handler) providerByID(providerID string) (map[string]any, bool) {
	target := strings.ToLower(strings.TrimSpace(providerID))
	for _, provider := range h.store.LoadAPIProviders() {
		if stringValue(provider["id"], "") == target {
			return provider, true
		}
	}
	return nil, false
}

func (h *Handler) GetComfyUIInstances(c *gin.Context) {
	c.JSON(http.StatusOK, gin.H{"instances": h.cfg.ComfyUIInstances})
}

func (h *Handler) firstPartyKeysResponse() gin.H {
	return gin.H{
		"has_openai_key":                h.cfg.OpenAIAPIKey != "",
		"openai_key_preview":            store.MaskSecret(h.cfg.OpenAIAPIKey),
		"has_volcengine_ark_key":        h.cfg.VolcengineArkAPIKey != "",
		"volcengine_ark_key_preview":    store.MaskSecret(h.cfg.VolcengineArkAPIKey),
		"has_volcengine_access_key":     h.cfg.VolcengineAccessKeyID != "",
		"volcengine_access_key_preview": store.MaskSecret(h.cfg.VolcengineAccessKeyID),
		"has_volcengine_secret_key":     h.cfg.VolcengineSecretAccessKey != "",
		"volcengine_secret_key_preview": store.MaskSecret(h.cfg.VolcengineSecretAccessKey),
		"volcengine_project_name":       h.cfg.VolcengineProjectName,
		"volcengine_region":             h.cfg.VolcengineRegion,
	}
}

func joinStringAny(value any) string {
	items, ok := value.([]any)
	if !ok {
		return ""
	}
	out := make([]string, 0, len(items))
	for _, item := range items {
		if text := stringValue(item, ""); text != "" {
			out = append(out, text)
		}
	}
	return strings.Join(out, ",")
}

func firstNonEmptyString(value, fallback string) string {
	if value != "" {
		return value
	}
	return fallback
}

func modelsEndpoint(baseURL string) string {
	if strings.HasSuffix(baseURL, "/v1") {
		return baseURL + "/models"
	}
	return baseURL + "/v1/models"
}

func getJSONWithBearer(endpoint, apiKey string, timeout time.Duration) (map[string]any, int, error) {
	req, err := http.NewRequest(http.MethodGet, endpoint, nil)
	if err != nil {
		return nil, 0, err
	}
	req.Header.Set("Authorization", "Bearer "+apiKey)
	req.Header.Set("Accept", "application/json")
	client := &http.Client{Timeout: timeout}
	resp, err := client.Do(req)
	if err != nil {
		return nil, 0, err
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(io.LimitReader(resp.Body, 2*1024*1024))
	if err != nil {
		return nil, resp.StatusCode, err
	}
	if len(body) == 0 {
		return map[string]any{}, resp.StatusCode, nil
	}
	var raw map[string]any
	if err := json.Unmarshal(body, &raw); err != nil {
		return map[string]any{"raw": string(body)}, resp.StatusCode, nil
	}
	return raw, resp.StatusCode, nil
}

func rawText(raw map[string]any) string {
	if raw == nil {
		return ""
	}
	if text := stringValue(raw["raw"], ""); text != "" {
		return text
	}
	data, err := json.Marshal(raw)
	if err != nil {
		return ""
	}
	return string(data)
}

func extractModelIDs(raw map[string]any) []string {
	var items any
	if raw != nil {
		items = raw["data"]
		if items == nil {
			items = raw["models"]
		}
		if items == nil {
			items = raw["list"]
		}
	}
	list, ok := items.([]any)
	if !ok {
		return []string{}
	}
	seen := map[string]bool{}
	out := make([]string, 0, len(list))
	for _, item := range list {
		id := ""
		if text, ok := item.(string); ok {
			id = text
		} else if obj, ok := item.(map[string]any); ok {
			id = firstNonEmptyString(stringValue(obj["id"], ""), firstNonEmptyString(stringValue(obj["name"], ""), stringValue(obj["model"], "")))
		}
		id = strings.TrimSpace(id)
		if id != "" && !seen[id] {
			seen[id] = true
			out = append(out, id)
		}
	}
	sort.Strings(out)
	return out
}

func groupModelIDs(ids []string) map[string][]string {
	grouped := map[string][]string{"image": {}, "chat": {}, "video": {}}
	for _, id := range ids {
		grouped[classifyModelID(id)] = append(grouped[classifyModelID(id)], id)
	}
	return grouped
}

func classifyModelID(modelID string) string {
	lc := strings.ToLower(modelID)
	videoKeys := []string{"veo", "sora", "wan2", "wanx", "doubao-seedance", "doubao-1", "kling", "hailuo", "video", "t2v-", "i2v-", "s2v"}
	for _, key := range videoKeys {
		if strings.Contains(lc, key) {
			return "video"
		}
	}
	imageKeys := []string{"image", "dalle", "dall-e", "imagen", "flux", "stable", "sdxl", "midjourney", "nano-banana", "ideogram", "fal-ai", "z-image", "qwen-image", "klein"}
	for _, key := range imageKeys {
		if strings.Contains(lc, key) {
			return "image"
		}
	}
	return "chat"
}

func truncate(value string, max int) string {
	runes := []rune(value)
	if len(runes) <= max {
		return value
	}
	return string(runes[:max])
}
